File: 1582-design-browser-history/design-browser-history.py
class Node():
    def __init__(self, page = ''):
        self.page = page
        self.next = None
        self.prev = None


class BrowserHistory(object):

    def __init__(self, homepage):
        """
        :type homepage: str
        """
        self.root = Node(homepage)
        

    def visit(self, url):
        """
        :type url: str
        :rtype: None
        """
        newPage = Node(url)

        self.root.next = newPage
        newPage.prev = self.root

        self.root = newPage

    def back(self, steps):
        """
        :type steps: int
        :rtype: str
        """
        i = 0
        while self.root and self.root.prev and i < steps:
            self.root = self.root.prev
            i += 1
        return self.root.page

    def forward(self, steps):
        """
        :type steps: int
        :rtype: str
        """
        i = 0
        while self.root and self.root.next and i < steps:
            self.root = self.root.next
            i += 1
        return self.root.page
        


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)

File: 1582-design-browser-history/design-browser-history.cpp
struct Node {
    string val;
    Node* prev=NULL;
    Node* next=NULL;
};
class BrowserHistory {
public:
Node* curr=NULL;
    BrowserHistory(string homepage) {
        curr=new Node();
        curr->val=homepage;
    }
    
    void visit(string url) {
        Node* temp=new Node();
        temp->val=url;
        curr->next=temp;
        temp->prev=curr;
        curr=temp;
    }
    
    string back(int steps) {
        while(curr->prev && steps>0)
        {
            curr=curr->prev;
            steps--;
        }
        return curr->val;
        
    }
    
    string forward(int steps) {
       while(curr->next&& steps>0)
       {
           curr=curr->next;
           steps--;
       } 
       return curr->val;
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */